### PR TITLE
Merge type hints from microsoft/python-type-stubs

### PR DESCRIPTION
**Pull request by Avasam**
# What does this PR do?

Merge type definitions from https://github.com/microsoft/python-type-stubs/tree/main/transformers-stubs so it can be removed from Pylance.
This is also work towards #16059 
I cross-checked the types with what I got at runtime.
I also ran `pyright --pythonversion=3.7` on both files to sanity check I'm not writing anything that will obviously break at runtime under Python 3.7

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section? (yes but make commands are not working on my machine)
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. (no)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation). (I guess? type hints are docs, docstrings and external docs should stay the same from my PR)
- [ ] Did you write any new necessary tests? (no, if you test using mypy/pyright this should already be picked up. Unit tests should naturally break if using syntax or imports incompatible with 3.7)

## Who can review?

🤷 The list below doesn't mention typing / type hints
I guess @Rocketknight1 who opened #16059

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
Thanks, but we are not interested in creating dependencies on the auto module over all those new modules just for the sake of type hinting. This makes Transformers suddenly have a hard dependency on tokenizers for instance, and the tokenizers imported for Fnet, T5 or Pegasus are wrong.

**Comment by Avasam** [comment]
> the tokenizers imported for Fnet, T5 or Pegasus are wrong

I copied exactly what I obtained from runtime inspection (see image below). And `FNetTokenizer`, `PegasusTokenizer`, `T5Tokenizer` and `RagTokenizer` have no common parent with `PreTrainedTokenizer`. Unless there's a `Protocol` I could use instead.
image

> not interested in creating dependencies on the auto module over all those new modules

Updated to not have runtime dependencies

**Comment by sgugger** [comment]
Python is not a statically typed language and your runtime inspection will be different form another user's runtime inspection depending on the packages installed. Again this is way more headache that what we want to deal with and the benefits of adding type hints, so we won't merge any type hints in the auto module.

File: src/transformers/models/auto/auto_factory.py [hunk 1]
```diff
@@ -570,7 +591,7 @@ def getattribute_from_module(module, attr):
         raise ValueError(f"Could not find {attr} in {transformers_module}!")
 
 
-class _LazyAutoMapping(OrderedDict):
+class _LazyAutoMapping(_LazyAutoMappingOrderedDict):
```
Comment (sgugger) [inline src/transformers/models/auto/auto_factory.py:594]: The base class should not be changed.

File: src/transformers/models/auto/auto_factory.py [hunk 2]
```diff
@@ -13,16 +13,37 @@
...
 
 
+if TYPE_CHECKING:
+    # Avoid runtime dependency on other modules for typing needs
+    from ...tokenization_utils_fast import PreTrainedTokenizerFast
+
+    _LazyAutoMappingValue = Tuple[
+        # Tokenizers will depend on packages installed, too much variance and there are no common base or Protocol
+        Type[Union[Any, None]],
+        Type[Union[PreTrainedTokenizerFast, None]],
```
Comment (sgugger) [inline src/transformers/models/auto/auto_factory.py:37]: The value can be a tokenizer, a config or a model, so the type annotation is wrong here.

File: src/transformers/models/auto/configuration_auto.py [hunk 3]
```diff
@@ -635,7 +649,7 @@ def config_class_to_model_type(config):
     return None
 
 
-class _LazyConfigMapping(OrderedDict):
+class _LazyConfigMapping(_LazyConfigMappingOrderedDict):
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:652]: Same, let's not change the base class here.

File: src/transformers/models/auto/configuration_auto.py [hunk 4]
```diff
@@ -13,17 +13,31 @@
...
 
+# Support Python 3.8 and below
+if TYPE_CHECKING:
+    _LazyLoadAllMappingsOrderedDict = OrderedDict[str, str]
+    _LazyConfigMappingOrderedDict = OrderedDict[str, Type[PretrainedConfig]]
+else:
+    _LazyLoadAllMappingsOrderedDict = OrderedDict
+    _LazyConfigMappingOrderedDict = OrderedDict
+
+_T = TypeVar("_T")
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:39]: We accept type annotations that make the code easier to read or understand. Those are not the kind that make the code easier to read.

File: src/transformers/models/auto/configuration_auto.py [hunk 5]
```diff
@@ -689,7 +703,7 @@ def register(self, key, value):
 CONFIG_MAPPING = _LazyConfigMapping(CONFIG_MAPPING_NAMES)
 
 
-class _LazyLoadAllMappings(OrderedDict):
+class _LazyLoadAllMappings(_LazyLoadAllMappingsOrderedDict):
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:706]: The base class should be left as is.

File: src/transformers/models/auto/configuration_auto.py [hunk 6]
```diff
@@ -787,8 +801,8 @@ def _list_model_options(indent, config_to_class=None, use_model_types=True):
     return "\n".join(lines)
 
 
-def replace_list_option_in_docstrings(config_to_class=None, use_model_types=True):
-    def docstring_decorator(fn):
+def replace_list_option_in_docstrings(config_to_class=None, use_model_types=True) -> Callable[[_T], _T]:
+    def docstring_decorator(fn: _T) -> _T:
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:805]: Those type annotations do not help understand the code and should be removed.

**Review by sgugger — COMMENTED** [review]
I've pointed out a bit more in the review what we look for in type annotations but we only use type annotations to make the code easier to read and we prefer to have no type annotations at all than wrong ones or ones that make the code harder to read.

**Comment by Avasam** [comment]
Since it's not possible to get accurate and useful inline generic type hints without changing the base class to an alias due to Python 3.8 support: To be reconsidered once python 3.8 support is dropped.

I'll backport this to https://github.com/microsoft/python-type-stubs so they're at least accurate. Which may or may not be migrated to typeshed at some point.

**Pull request closed without merging**